lib/Target/Lattigo/LattigoEmitter.cpp

```diff
@@ -0,0 +1,312 @@
+                                op.getPlaintext(), "Encode", false);
+}
+
+LogicalResult LattigoEmitter::printOperation(BGVDecodeOp op) {
+  return printEvalInplaceMethod(op.getDecoded(), op.getEncoder(),
+                                op.getPlaintext(), op.getValue(), "Decode",
+                                false);
+}
+
+LogicalResult LattigoEmitter::printOperation(BGVAddOp op) {
+  return printEvalNewMethod(op.getResult(), op.getEvaluator(),
+                            {op.getLhs(), op.getRhs()}, "AddNew", true);
+}
+
+LogicalResult LattigoEmitter::printOperation(BGVSubOp op) {
+  return printEvalNewMethod(op.getResult(), op.getEvaluator(),
+                            {op.getLhs(), op.getRhs()}, "SubNew", true);
+}
+
+LogicalResult LattigoEmitter::printOperation(BGVMulOp op) {
+  return printEvalNewMethod(op.getResult(), op.getEvaluator(),
+                            {op.getLhs(), op.getRhs()}, "MulNew", true);
+}
+
+std::string printDenseI32ArrayAttr(DenseI32ArrayAttr attr) {
+  std::string res = "[]int{";
+  res += commaSeparated(attr.asArrayRef());
+  res += "}";
+  return res;
+}
+
+LogicalResult LattigoEmitter::printOperation(BGVNewParametersFromLiteralOp op) {
+  std::string errResult = ", err";
+  os << getName(op.getResult()) << errResult
+     << " := bgv.NewParametersFromLiteral(";
+  os << "bgv.ParametersLiteral{";
+  os.indent();
+  os << "LogN: " << op.getParamsLiteral().getLogN() << ",\n";
+  os << "LogQ: " << printDenseI32ArrayAttr(op.getParamsLiteral().getLogQ())
+     << ",\n";
+  os << "LogP: " << printDenseI32ArrayAttr(op.getParamsLiteral().getLogP())
+     << ",\n";
+  os << "PlaintextModulus: " << op.getParamsLiteral().getPlaintextModulus()
+     << ",\n";
+  os.unindent();
+  os << "})\n";
+  printErrPanic();
+  return success();
+}
+
+void LattigoEmitter::printErrPanic() {
+  os << "if err != nil {\n";
+  os.indent();
+  os << "panic(err)\n";
+  os.unindent();
+  os << "}\n";
+}
+
+LogicalResult LattigoEmitter::printNewMethod(::mlir::Value result,
+                                             ::mlir::ValueRange operands,
+                                             std::string_view op, bool err) {
+  std::string errResult = err ? ", err" : "";
+  os << getName(result);
+  os << errResult << " := " << op << "(";
+  os << getCommaSeparatedNames(operands);
+  os << ")\n";
+  if (err) {
+    printErrPanic();
+  }
+  return success();
+}
+
+LogicalResult LattigoEmitter::printEvalInplaceMethod(
+    ::mlir::Value result, ::mlir::Value evaluator, ::mlir::Value operand,
+    ::mlir::Value operandInplace, std::string_view op, bool err) {
+  std::string errResult = err ? ", err := " : "";
+  os << errResult << getName(evaluator) << "." << op << "(" << getName(operand)
+     << ", " << getName(operandInplace) << ");\n";
+  if (err) {
+    printErrPanic();
+  }
+  // workaround
+  os << getName(result) << " := " << getName(operandInplace) << "\n";
```

Comment:
To your question in the PR notes, I would say this is fine. My only concern is that I recall Golang performs a copy on assignment. This would be acceptable if we knew that all the values being copied were just pointers, potentially unacceptable if it were copying entire ciphertexts. I don't know lattigo well enough to say, can we confirm and possibly add a code comment with the confirmation?

Reply:
In lattigo all function deals with pointers, so I think it is acceptable

```
package bgv

func NewPlaintext(params Parameters, level int) (pt *rlwe.Plaintext) {

func NewCiphertext(params Parameters, degree, level int) (ct *rlwe.Ciphertext) {
```